rpcmiddleware/interface.go

```diff
@@ -0,0 +1,345 @@
+
+	// DefaultInterceptTimeout is the default maximum time we're allowed to
+	// take to respond to an RPC middleware interception request.
+	DefaultInterceptTimeout = 3 * time.Second
+
+	// acceptRequestHandler is a RequestHandler that accepts all requests.
+	acceptRequestHandler requestHandler = func(proto.Message) error {
+		return nil
+	}
+
+	// passThroughResponseHandler is a responseHandler that does not modify
+	// the response and just passes it through.
+	passThroughResponseHandler responseHandler = func(
+		proto.Message) (proto.Message, error) {
+
+		return nil, nil
+	}
+)
+
+// RequestInterceptor is a type that can intercept an RPC request.
+type RequestInterceptor interface {
+	// Name returns the name of the interceptor.
+	Name() string
+
+	// ReadOnly returns true if this interceptor should be registered in
+	// read-only mode. In read-only mode no custom caveat name can be
+	// specified.
+	ReadOnly() bool
+
+	// CustomCaveatName returns the name of the custom caveat that is
+	// expected to be handled by this interceptor. Cannot be specified in
+	// read-only mode.
+	CustomCaveatName() string
+
+	// Intercept processes an RPC middleware interception request and
+	// returns the interception result which either accepts or rejects the
+	// intercepted message.
+	Intercept(context.Context,
+		*lnrpc.RPCMiddlewareRequest) (*lnrpc.RPCMiddlewareResponse,
+		error)
+}
+
+// requestHandler is a function type for a generic gRPC request message handler
+// that can accept (=return nil) or refuse (=return non-nil error with rejection
+// reason) an incoming request.
+type requestHandler func(req proto.Message) error
+
+// responseHandler is a function type for a generic gRPC response message
+// handler that can pass through the response (=return nil, nil), replace the
+// response with a new message of the same type (=return non-nil message, nil
+// error) or abort the call by returning a non-nil error.
+type responseHandler func(req proto.Message) (proto.Message, error)
+
+// RoundTripChecker is a type that represents a basic request/response round
+// trip checker.
+type RoundTripChecker interface {
+	// AcceptsRequest returns true if the checker accepts protobuf request
+	// messages of the given type. This is mainly a safety feature to make
+	// sure a round trip checker is implemented correctly.
+	AcceptsRequest(protoreflect.MessageType) bool
+
+	// HandlesResponse returns true if the checker can handle protobuf
+	// response messages of the given type. This is mainly a safety feature
+	// to make sure a round trip checker is implemented correctly.
+	HandlesResponse(protoreflect.MessageType) bool
+
+	// HandleRequest is called for each incoming gRPC request message of the
+	// type declared to be accepted by AcceptsRequest. The handler can
+	// accept (=return nil) or refuse (=return non-nil error with rejection
+	// reason) an incoming request.
+	HandleRequest(proto.Message) error
```

Comment:
could there maybe be a situation where we want to edit the request before it goes through? (like we do for the response)